homeassistant/components/light/group.py

```diff
@@ -0,0 +1,262 @@
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def state(self) -> str:
+        """Return the state."""
+        return self._state
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self.state == STATE_ON
+
+    @property
+    def brightness(self) -> Optional[int]:
+        """Return the brightness of this light between 0..255."""
+        return self._brightness
+
+    @property
+    def xy_color(self) -> Optional[Tuple[float, float]]:
+        """Return the XY color value [float, float]."""
+        return self._xy_color
+
+    @property
+    def rgb_color(self) -> Optional[Tuple[int, int, int]]:
+        """Return the RGB color value [int, int, int]."""
+        return self._rgb_color
+
+    @property
+    def color_temp(self) -> Optional[int]:
+        """Return the CT color value in mireds."""
+        return self._color_temp
+
+    @property
+    def min_mireds(self) -> Optional[int]:
+        """Return the coldest color_temp that this light supports."""
+        return self._min_mireds
+
+    @property
+    def max_mireds(self) -> Optional[int]:
+        """Return the warmest color_temp that this light supports."""
+        return self._max_mireds
+
+    @property
+    def white_value(self) -> Optional[int]:
+        """Return the white value of this light between 0..255."""
+        return self._white_value
+
+    @property
+    def effect_list(self) -> Optional[List[str]]:
+        """Return the list of supported effects."""
+        return self._effect_list
+
+    @property
+    def effect(self) -> Optional[str]:
+        """Return the current effect."""
+        return self._effect
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed for a group light."""
+        return False
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Forward the turn_on command to all lights in the group."""
+        for entity_id in self._entity_ids:
+            kwargs[ATTR_ENTITY_ID] = entity_id
+            yield from self.hass.services.async_call(
+                'light', SERVICE_TURN_ON, kwargs, blocking=True)
+
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs):
+        """Forward the turn_off command to all lights in the group."""
+        for entity_id in self._entity_ids:
+            kwargs[ATTR_ENTITY_ID] = entity_id
+            yield from self.hass.services.async_call(
+                'light', SERVICE_TURN_OFF, kwargs, blocking=True)
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Query all members and determine the group state."""
+        states = self._child_states()
+
+        self._state = _determine_on_off_state(states)
```

Comment:
All these functions will do their own iteration over all items in the list. That is very inefficient.

```python
states = []
min_mireds = 154

for entity_id in self._entity_ids:
    state = self.hass.states.get(entity_id)
    if state is None:
        continue

    # In case you still need to map over all states for some checks
    states.append(state)

    # Etc…
    min_mireds = min(min_mireds, state.attributes[MIN_MIREDS])
```

Reply:
Nice thing of doing it like this is that you can also not do a bunch of checks when the state is off.

Reply:
> All these functions will do their own iteration over all items in the list.

Technically they only go through the iteration until they find a state that matches - that's also why I'm using the generator syntax and `next(..., default)` in `_reduce_attribute`. Next, complexity-wise isn't having a global `for-loop` over all states just the same as having a local `for-loop` for all attributes? I'm not a python expert, but I mean maybe python has *a bit* of overhead with starting for-loops, but would that tiny speed benefit really be worth it? I mean this function will probably not be called too often.

> min_mireds = min(min_mireds, state.attributes[MIN_MIREDS])

I mean if I understand correctly this kinda touches on https://github.com/home-assistant/architecture/issues/13; do we want to report the first light state or the union of all supported features/states?

Reply:
The example with `min_mireds` is quite, but with `brightness` for example where one has to handle `None` I've come up with these two solutions with using average:

```python
brightnesses = []

for state in states:
    if state.state != STATE_ON:
        continue

    if state.attributes.get('brightness') is not None:
        brightnesses.append(state.attributes['brightness'])
        
if brightnesses:
    self._brightness = sum(brightnesses) / len(brightnesses)
else:
    self._brightness = None
```

```python
self._brightness = None
num_brightness = 0

for state in states:
    if state.state != STATE_ON:
        continue

    if state.attributes.get('brightness') is not None:
        self._brightness = (self.brightness or 0) + state.attributes['brightness']
        num_brightness += 1

if num_brightness > 0:
    self._brightness = self._brightness / num_brightness
```

I mean one could probably also design helper methods around these statements, but is the speed benefit really worth the messier code? Also: at least some of the slowness of the previous code has now been fixed by pre-calculating a list of states that are `on`.

Reply:
Ok, I just ran a speed test for both methods and the current method with using a helper actually seems faster: (Scroll to the bottom; `f` is current method, `g` would be *new*  method)

https://gist.github.com/OttoWinter/58ce0fd4c7743d233af8a10ed7ff940a

(I might be doing something wrong though.)